### PR TITLE
Versioning_Engine: method to easily generate method keys for versioning

### PR DESCRIPTION
**Pull request by adecler**
### Issues addressed by this PR
Provide helper method/component to go along with this PR: https://github.com/BHoM/BHoM_UI/pull/247

image

This makes it way easier to populate the inputs of a `PreviousVersion` attribute.

### CODE REVIEW

File: Versioning_Engine/Versioning_Engine.csproj [hunk 1]
```diff
@@ -37,6 +37,10 @@
     <Reference Include="MongoDB.Bson, Version=2.4.4.0, Culture=neutral, processorArchitecture=MSIL">
       <HintPath>..\packages\MongoDB.Bson.2.4.4\lib\net45\MongoDB.Bson.dll</HintPath>
     </Reference>
+    <Reference Include="Reflection_oM, Version=3.0.0.0, Culture=neutral, processorArchitecture=MSIL">
+      <SpecificVersion>False</SpecificVersion>
+      <HintPath>..\..\BHoM\Build\Reflection_oM.dll</HintPath>
+    </Reference>
```
Comment (FraserGreenroyd) [inline Versioning_Engine/Versioning_Engine.csproj:43]: Reference should not have version, culture, or processor architecture and should be set to `copy local` false

File: Versioning_Engine/Compute/VersionKey.cs [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    public static partial class Compute
+    {
+        /***************************************************/
+        /**** Public Methods                            ****/
+        /***************************************************/
+
+        [Description("Provide a string representation of a method as it used for versioning by the PreviousVersion attribute.")]
+        [Input("declaringType", "Type in which the method is declared. You can use just the name of the type or include a (part of the) namespace in front of it.")]
+        [Input("methodName", "Name of the method. It has to be the exact string")]
+        [Output("Keys", "String representation for each method that matches the input filters.")]
```
Comment (FraserGreenroyd) [inline Versioning_Engine/Compute/VersionKey.cs:52]: I think consistency we use lowercase first letters on output names as with input names

File: Versioning_Engine/Query/VersionKey.cs [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+{
+    public static partial class Query
+    {
+        /***************************************************/
+        /**** Public Methods                            ****/
+        /***************************************************/
+
+        [Description("Provide a string representation of a method as it used for versioning by the PreviousVersion attribute.")]
+        [Input("methodName", "Method to ge tthe key from")]
+        [Output("Key", "String representation of the method as it will be used by the PreviousVersion attribute.")]
```
Comment (FraserGreenroyd) [inline Versioning_Engine/Query/VersionKey.cs:51]: I think consistency we use lowercase first letters on output names as with input names

File: Versioning_Engine/Compute/VersionKey.cs [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+{
+    public static partial class Compute
+    {
+        /***************************************************/
+        /**** Public Methods                            ****/
+        /***************************************************/
+
+        [Description("Provide a string representation of a method as it used for versioning by the PreviousVersion attribute.")]
+        [Input("declaringType", "Type in which the method is declared. You can use just the name of the type or include a (part of the) namespace in front of it.")]
+        [Input("methodName", "Name of the method. It has to be the exact string")]
```
Comment (FraserGreenroyd) [inline Versioning_Engine/Compute/VersionKey.cs:51]: Looking at the code, it appears we can have a blank `methodName` to use `.ctor` options - is it worth including this in the description

File: Versioning_Engine/Query/VersionKey.cs [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+namespace BH.Engine.Versioning
+{
+    public static partial class Query
+    {
+        /***************************************************/
+        /**** Public Methods                            ****/
+        /***************************************************/
+
+        [Description("Provide a string representation of a method as it used for versioning by the PreviousVersion attribute.")]
+        [Input("methodName", "Method to ge tthe key from")]
```
Comment (FraserGreenroyd) [inline Versioning_Engine/Query/VersionKey.cs:50]: Should probably be:
```suggestion
        [Input("methodName", "Method to get the key from")]
```

**Comment by FraserGreenroyd** [comment]
Looks like everything is working as described for me from this code :smile:

image

**Comment by adecler** [comment]
@FraserGreenroyd , I fixed the problems raised in your review

**Comment by kThorsager** [comment]
Perhaps intended due to `Modify` -> `Compute` not working but if a method is found in both it returns nothing.
image

Worth flagging as that is a planned feature ([think this one](https://github.com/BHoM/Versioning_Toolkit/issues/51))

**Comment by FraserGreenroyd** [comment]
> Perhaps intended due to `Modify` -> `Compute` not working but if a method is found in both it returns nothing.

I think from the demo @adecler gave this morning, the matching starts from the right and works left, so with those inputs it would be looking for `BH.Engine.Geometry.Join` which of course doesn't exist (based on my understanding of it).

Maybe a warning to the user to say that the provided thing couldn't be found might be worthwhile? Or if the issue you link is the right one then solving that at a later date.

**Comment by kThorsager** [comment]
> > Perhaps intended due to `Modify` -> `Compute` not working but if a method is found in both it returns nothing.
> 
> I think from the demo @adecler gave this morning, the matching starts from the right and works left, so with those inputs it would be looking for `BH.Engine.Geometry.Join` which of course doesn't exist (based on my understanding of it).
> 
> Maybe a warning to the user to say that the provided thing couldn't be found might be worthwhile? Or if the issue you link is the right one then solving that at a later date.

No, sorry, re-read the description and sounds right (but thought the demo showed using just a part of the namespace)

**Comment by FraserGreenroyd** [comment]
> No, sorry, re-read the description and sounds right (but thought the demo showed using just a part of the namespace)

It does, but it started from the right rather than the left. So I think something like `Modify` and `Join` would find the right one (or all the ones that match) because there is `*.Modify.Join` methods if that makes sense?

**Comment by adecler** [comment]
A simple way to settle that discussion is to look at the code:
```c#
return Engine.Reflection.Query.AllMethodList()
                .Where(x => x.Name == methodName && x.DeclaringType.FullName.EndsWith(declaringType))
                .Select(x => x.VersioningKey())
                .ToList();
```

I was using `EndsWith` instead of `==` so that users wouldn't have to figure out whether to provide just the name of the declaring type or its full name (including namespace). The nice side effect was that yo could even do partial namespaces. I do however regret showing you that since it has been misinterpreted. The input name is `declaringType`. As far as I know, `BH.Engine.Geometry` is a namespace, not a declaring type.

**Pull request merged**